pxd_bio_blkmq.c

```diff
@@ -580,19 +578,8 @@ static void fp_handle_specialops(struct work_struct *work) {
         bdev = get_bdev(file);
         q = bdev_get_queue(bdev);
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 8, 0) || defined(REQ_PREFLUSH)
-        if (REQ_OP(rq) == REQ_OP_DISCARD) {
-                atomic_inc(&pxd_dev->fp.nio_discard);
-        } else {
-                BUG_ON("unexpected condition");
-        }
-#else
-        if (REQ_OP(rq) & REQ_DISCARD) {
-                atomic_inc(&pxd_dev->fp.nio_discard);
-        } else {
-                BUG_ON("unexpected condition");
-        }
-#endif
+        BUG_ON(!rq_is_special(rq));
```

Comment:
rq_is_special() only checks for discard, Based on the below it should also check for WRITE_SAME and zeroout if there's an op for it.

Reply:
No. px device only supports read/write and discard, write same and write zeros are not supported on px device.
The below logic converts the received discard to map to the backing device (btrfs file or dmthin volume).
It may convert discard internally to write same or write zero if discard is not available.